modules/core/src/main/java/org/apache/ignite/internal/jdbc2/lob/JdbcBlobBuffer.java

```diff
@@ -0,0 +1,354 @@
+
+/**
+ * Buffer storing the binary data.
+ *
+ * <p>Buffer can start working in read-only mode if created wrapping the existing byte array which
+ * can not be modified. Any write operation switches it lazily to the read-write mode. This allows
+ * to prevent the unnecessary data copying.
+ *
+ * <p>Data is read via the InputStream API and modified via the OutputStream one. Changes done via
+ * OutputStream are visible via the InputStream even if InputStream is created before changes done.
+ *
+ * <p>InputStream and OutputStream created remain valid even if the underlying data storage changed from
+ * read-only to read-write.
+ *
+ * <p>Note however that implementation is not thread-safe.
+ */
+public class JdbcBlobBuffer {
+    /** The underlying data storage. */
+    private JdbcBlobStorage storage;
+
+    /**
+     * Create buffer which wraps the existing byte array and start working in the read-only mode.
+     *
+     * @param arr The byte array to be wrapped.
+     * @param off The offset to the first byte to be wrapped.
+     * @param len The length in bytes of the data to be wrapped.
+     */
+    public static JdbcBlobBuffer createReadOnly(byte[] arr, int off, int len) {
+        return new JdbcBlobBuffer(new JdbcBlobReadOnlyMemoryStorage(arr, off, len));
+    }
+
+    /**
+     * Create buffer which takes ownerhip of and wraps the existing byte array and starts working in
+     * the read-write mode.
+     *
+     * @param arr The byte array to be wrapped.
+     */
+    public static JdbcBlobBuffer createReadWrite(byte[] arr) {
+        return new JdbcBlobBuffer(new JdbcBlobMemoryStorage(arr));
+    }
+
+    /**
+     * Create empty buffer which starts working in the read-write mode.
+     */
+    public static JdbcBlobBuffer createReadWrite() {
+        return new JdbcBlobBuffer(new JdbcBlobMemoryStorage());
+    }
+
+    /**
+     * Create shallow copy of the buffer passed.
+     *
+     * <p>Sharing of the underlying storage is intended.
+     *
+     * @param other Other buffer.
+     */
+    public static JdbcBlobBuffer shallowCopy(JdbcBlobBuffer other) {
+        return new JdbcBlobBuffer(other.storage);
+    }
+
+    /**
+     * Create buffer which wraps the passed storage instance.
+     * @param storage Storage instance.
+     */
+    private JdbcBlobBuffer(JdbcBlobStorage storage) {
+        this.storage = storage;
+    }
+
+    /**
+     * @return Total number of bytes in the buffer.
+     */
+    public long totalCnt() {
+        return storage.totalCnt();
+    }
+
+    /**
+     * Provide OutputStream through which the data can be written to buffer starting from
+     * the (zero-based) position.
+     *
+     * @param pos The zero-based offset to the first byte to be written. Must not be negative
+     *            or greater than total count of bytes in buffer.
+     *
+     * @return OutputStream instance.
+     */
+    public OutputStream getOutputStream(long pos) {
+        long totalCnt = storage.totalCnt();
```

Comment:
Looks like sanity checks double. Here and in

`JdbcBlob#setBytes`, `JdbcBlob#setBinaryStream` where it's used.
Can we keep only one copy of sanity checks?